src/ophyd_async/tango/base_devices/_base_device.py

```diff
@@ -0,0 +1,122 @@
+    proxy: Optional[Union[AsyncDeviceProxy, SyncDeviceProxy]] = None
+    _polling: tuple = (False, 0.1, None, 0.1)
+
+    def __init__(
+        self,
+        trl: Optional[str] = None,
+        device_proxy: Optional[Union[AsyncDeviceProxy, SyncDeviceProxy]] = None,
+        name: str = "",
+    ) -> None:
+        if not trl and not device_proxy:
+            raise ValueError("Either 'trl' or 'device_proxy' must be provided.")
+
+        self.trl = trl if trl else ""
+        self.proxy = device_proxy
+
+        self.create_children_from_annotations()
+        super().__init__(name=name)
+
+    async def connect(
+        self,
+        mock: bool = False,
+        timeout: float = DEFAULT_TIMEOUT,
+        force_reconnect: bool = False,
+    ):
+        async def closure():
+            try:
+                if self.proxy is None:
+                    self.proxy = await AsyncDeviceProxy(self.trl)
+            except Exception as e:
+                raise RuntimeError("Could not connect to device proxy") from e
+            return self
+
+        if self.trl in ["", None]:
+            self.trl = self.proxy.name()
+
+        await closure()
+        self.register_signals()
+        # set_name should be called again to propagate the new signal names
+        self.set_name(self.name)
+
+        await super().connect(mock=mock, timeout=timeout)
+
+    def register_signals(self):
+        annots = get_type_hints(self.__class__)
+        for name, obj_type in annots.items():
+            if hasattr(self, name):
+                signal = getattr(self, name)
+                if issubclass(type(signal), Signal):
+                    tango_name = name.lstrip("_")
+                    read_trl = f"{self.trl}/{tango_name}"
+                    datatype = infer_python_type(read_trl)
+                    backend = make_backend(
+                        datatype=datatype,
+                        read_trl=read_trl,
+                        write_trl=read_trl,
+                        device_proxy=self.proxy,
+                    )
+                    if self._polling[0]:
+                        backend.allow_events(False)
+                        backend.set_polling(*self._polling)
+                    signal._backend = backend  # noqa: SLF001
+
+    def create_children_from_annotations(self):
+        annots = get_type_hints(self.__class__)
+        for attr_name, obj_type in annots.items():
+            if isinstance(obj_type, type):
+                if obj_type is Signal:
+                    tango_name = attr_name.lstrip("_")
+                    trl = f"{self.trl}/{tango_name}"
+                    setattr(
+                        self, attr_name, infer_signal_frontend(trl=trl, name=attr_name)
+                    )
+                elif issubclass(obj_type, Signal):
+                    setattr(self, attr_name, obj_type(name=attr_name))
+
+
+# --------------------------------------------------------------------
+def tango_polling(*args):
```

Comment:
I'm tempted to suggest that polling be per-signal rather than per-device, partly for consistency and partly for flexibility (an ophyd-async device comprising signals from multiple tango servers), but I'm too ignorant about tango servers and common use cases for that to be a firm suggestion.

Reply:
I am not certain on this either. In general, Tango servers can publish events so polling isn't necessary. But some facilities (like DESY) do not use Tango events. Many fast polling signals might effectively DDOS the tango server so one might need to be careful not to do that. The decorator can wrap any class to set the polling attribute so it should work at the device level, or the signal backend. If not, its because I have been inconsistent in how I define the polling attribute. I'll make sure it can work at the device or signal level.

Reply:
I changed tango_polling to enable either device level with a tuple, signal level polling with a dict, or both.